File: Trees/230. Kth Smallest Element in a BST/Solution.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        # N : no of node in root
        # Time Complexity: O(N)
        # Space Compleixty: O(N) -> Stack Space
        kthVal = 0
        def dfs(root):
            nonlocal kthVal, k
            if not root:
                return
            
            dfs(root.left)

            k -= 1
            if (k == 0):
                kthVal = root.val
                return

            dfs(root.right)
            return
        dfs(root)
        return kthVal

File: Trees/230. Kth Smallest Element in a BST/Solution.java
//Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution {
    private int kthVal;
    private int kth;
    private void dfs(TreeNode root) {
        if (root == null) return;
        dfs(root.left);
        kth--;
        if (kth == 0) kthVal = root.val;
        dfs(root.right);
        return;
    }
    public int kthSmallest(TreeNode root, int k) {
        // N : no of node in root
        // Time Complexity: O(N)
        // Space Compleixty: O(N) -> Stack Space
        kth = k;
        dfs(root);
        return kthVal;
    }
}